### ISSUE TITLE
Slack RTM typing indicator

### ISSUE THREAD

**Issue by crocodele** [mpociot/botman#264]
The Slack RTM driver currently has an empty `types()` method. I've tried to implement the typing indicator using the RTM API ([documentation](https://api.slack.com/rtm#typing_indicators)). It works in the sense that the typing indicator is shown, but for some reason the indicator appears only after the reply has been sent. The result is the same regardless of the sleep time (I tried with 10 seconds and 1 second). Do you have any ideas why showing the typing indicator is deferred like this? Is it related to websocket calls vs. HTTP API calls?

**slack-client / RealTimeClient.php:**

```
public function setAsTyping(ChannelInterface $channel)
{
	if (!$this->connected) {
		return Promise\reject(new ConnectionException('Client not connected. Did you forget to call `connect()`?'));
	}

	$data = [
		'id' => ++$this->lastMessageId,
		'type' => 'typing',
		'channel' => $channel->data['id'],
	];
	$this->websocket->send(json_encode($data));
}
```

**botman / SlackRTMDriver.php:**

```
public function types(Message $matchingMessage)
{
	$channel = null;

	$this->getChannel($matchingMessage)->then(function ($_channel) use (&$channel) {
		$channel = $_channel;
	});

	if (is_null($channel)) {
		$this->client->getDMById($matchingMessage->getChannel())->then(function ($_channel) use (&$channel) {
		   $channel = $_channel;
		});
	}

	if (! is_null($channel)) {
		$this->client->setAsTyping($channel);
	}
}
```

**test:**

```
$loop = Factory::create();
$config = ['slack_token' => 'SLACK_TOKEN_HERE'];
$botMan = BotManFactory::createForRTM($config, $loop);

$botMan->hears('ping', function(BotMan $bot) {

    // Simulate typing
    $bot->typesAndWaits(1);

    // Respond
    $bot->reply('Pong');
});

$loop->run();
```

Thanks for an awesome library! 😎

**Comment by mpociot** [comment]
Hey,

yeah - I was looking into this quite a few times, but haven't figured it out yet.
This somehow has to do with the way the React PHP promises get called / the asynchronous websocket call :/

I could need some help from a React PHP pro on this one.

**Comment by crocodele** [comment]
OK, thanks. Perhaps the websocket message is sent on the next tick, which doesn't happen until after all the synchronous code has finished executing? I might at some point try to investigate what's going on, but I haven't used React PHP before so someone else is surely better equipped to have a go at it.

**Comment by crocodele** [comment]
I tried a few more things and got this working properly by adding `$this->loop->tick();` to the end of the `setAsTyping()` method above. However, manually ticking the event loop like this is considered a hack, and [`tick()` has been deprecated](https://github.com/reactphp/event-loop/pull/72) in the yet-to-be-released version 0.5.0 of ReactPHP's event loop component.

Not really sure if or how we can implement this by the book without making major changes in Botman. =/

**Comment by mpociot** [comment]
@crocodele thanks for looking into this!
Could you still provide a pull request for this?

I am not sure how and when we will adopt to React PHP 0.5.0, because the slack-client package is currently using `0.4.*` and things will probably break with 0.5.0 :)

So as long as it works for now - let's just go for it! 👍
